presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/rtree/Flatbush.java

```diff
@@ -0,0 +1,320 @@
+        checkArgument(degree > 0, "degree must be positive");
+        this.degree = degree;
+        this.items = requireNonNull(items, "items is null");
+        this.levelOffsets = Flatbush.calculateLevelIndices(items.length, degree);
+        this.tree = this.buildTree();
+    }
+
+    public Flatbush(T[] items)
+    {
+        this(items, DEFAULT_DEGREE);
+    }
+
+    /* Calculate the indices for each level.
+     *
+     * A given level contains a certain number of items.  We give it a capacity
+     * equal to the next multiple of `degree`, so that each parent will have
+     * an equal number (`degree`) of children.  This means the next level will
+     * have `capacity/degree` nodes, which yields a capacity equal to the next
+     * multiple, and so on, until the number of items is 1.
+     *
+     * Since we are storing each node as 4 doubles, we will actually multiply
+     * every index by 4.
+     */
+    private static int[] calculateLevelIndices(int numItems, int degree)
+    {
+        List<Integer> levelIndices = new ArrayList<Integer>();
+        // Leaf nodes start at 0, root is the last element.
+        levelIndices.add(Integer.valueOf(0));
+        int capacity;
+        int level = 0;
+        while (numItems > 1) {
+            capacity = Flatbush.findNextMultiple(numItems, degree);
+            levelIndices.add(levelIndices.get(level) + ENVELOPE_SIZE * capacity);
+            numItems = capacity / degree;
+            level += 1;
+        }
+        return levelIndices.stream().mapToInt(Integer::intValue).toArray();
+    }
+
+    /**
+     * Find the least multiple of `factor` gte to `n`.
+     * <p>
+     * n must be positive.
+     * TODO: Probably put this in utilities somewhere?
+     */
+    private static int findNextMultiple(int n, int factor)
+    {
+        return ((n + factor - 1) / factor) * factor;
+    }
+
+    public int getDegree()
+    {
+        return degree;
+    }
+
+    public T[] getItems()
+    {
+        return items;
+    }
+
+    public int getHeight()
+    {
+        return levelOffsets.length;
+    }
+
+    public boolean isEmpty()
+    {
+        return items.length == 0;
+    }
+
+    private DoubleBigArray buildTree()
+    {
+        // We initialize it to NaN, because all comparisons with NaN are false.
+        // Thus the normal intersection logic will not select uninitialized
+        // nodes.
+        DoubleBigArray tree = new DoubleBigArray(Double.NaN);
+        tree.ensureCapacity(levelOffsets[levelOffsets.length - 1] + ENVELOPE_SIZE);
+
+        sortByHilberIndex(this.items);
+
+        int writeOffset = 0;
+        int readOffset;
+        for (T item : items) {
+            tree.set(writeOffset++, item.getExtent().getXMin());
+            tree.set(writeOffset++, item.getExtent().getYMin());
+            tree.set(writeOffset++, item.getExtent().getXMax());
+            tree.set(writeOffset++, item.getExtent().getYMax());
+        }
+
+        int numChildren = items.length;
```

Comment:
this is the number of leaf nodes, right? if so, perhaps, rename to numLeaves or numItems

Reply:
It starts off as the number of leaf nodes, but in the next iteration of the loop it will be the number of parents of the leaf nodes, which are themselves the children of their parents (the next level of nodes).